teradata/datadog_checks/teradata/check.py

```diff
@@ -0,0 +1,168 @@
+# (C) Datadog, Inc. 2022-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+import json
+import time
+from contextlib import closing, contextmanager
+
+try:
+    import teradatasql
+
+    TERADATASQL_IMPORT_ERROR = None
+except ImportError as e:
+    teradatasql = None
+    TERADATASQL_IMPORT_ERROR = e
+
+
+from datadog_checks.base import AgentCheck, is_affirmative
+from datadog_checks.base.constants import ServiceCheck
+from datadog_checks.base.utils.db import QueryManager
+
+from .config_models import ConfigMixin
+from .queries import COLLECT_RES_USAGE, DEFAULT_QUERIES
+
+SERVICE_CHECK_CONNECT = 'can_connect'
+SERVICE_CHECK_QUERY = 'can_query'
+
+
+class TeradataCheck(AgentCheck, ConfigMixin):
+    __NAMESPACE__ = 'teradata'
+
+    def __init__(self, name, init_config, instances):
+        super(TeradataCheck, self).__init__(name, init_config, instances)
+
+        self._connect_params = None
+        self._connection = None
+
+        self._tags = [
+            'teradata_server:{}'.format(self.instance.get('server')),
+            'teradata_port:{}'.format(self.instance.get('port', 1025)),
+        ]
+        self._tags.extend(self.instance.get('tags', []))
+
+        manager_queries = []
+        if is_affirmative(self.instance.get('collect_res_usage', False)):
+            manager_queries.extend(COLLECT_RES_USAGE)
+        else:
+            manager_queries.extend(DEFAULT_QUERIES)
+
+        self._query_manager = QueryManager(
+            self,
+            self._execute_query_raw,
+            queries=manager_queries,
+            tags=self._tags,
+            error_handler=self._executor_error_handler,
+        )
+        self.check_initializations.append(self.initialize_config)
+        self.check_initializations.append(self._query_manager.compile_queries)
+
+        self._query_errors = 0
+
+    def check(self, _):
+        with self.connect() as conn:
+            if conn:
+                self._connection = conn
+                self._query_manager.execute()
+
+        self.submit_health_checks()
```

Comment:
Is this submitting a healthy connection service check regardless of what happens before?

Reply:
That's a good question. The check assumes that a healthy connection was made before calling the `submit_health_checks()` method . If the `connect()` method is unable to connect, it raises an exception, which prevents the `submit_health_checks()` method from getting called. Is that ok practice or should we ensure that the main `check()` method runs its code to the end?

Since the main purpose of `submit_health_checks()` is really to handle the `can_query` service check, do you think I should handle each service check separately? I.e. change `submit_health_checks()` => `submit_query_health_check()` to handle only `can_query` and then handle `can_connect` in the main `check()` method.

I've also added a try/except to main `check()` method catch any connection exceptions, removed the default OK status for `can_query`, and made sure that the `._query_errors` counter is reset on each check run